Use the custom @objc name in the available attribute when generates ObjC headers

pitiphong-p opened the pull request:
<!-- What's in this pull request? -->
Printing the custom `@objc` name in availability attributes in the generated Objective-C header files instead of printing the Swift name. This should help the compiler/fix-it to change the deprecated/unavailable API usage with the appropriate name

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-8231](https://bugs.swift.org/browse/SR-8231).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

pitiphong-p committed:
Use the custom @objc name in the Availablity attributes when generates the ObjC header

pitiphong-p committed:
Added: More test cases for the PrintAsObjC availability

pitiphong-p commented:
I have a few questions for this PR before proceeding
1. Should I open this PR to the swift-4.2 branch instead?
1. I still don't know how to get a Decl for a class given by a name so I left the work for the case of a class and will work on it as soon as I know a way (https://github.com/apple/swift/pull/18081/files#diff-27516adf920dd85ca69f0e537e6913e6R1910)
1. I don't sure how we should handle a case where the developer has a whole class to replace another class and they mark the old class's members with a rename availability attribute (You can see an example here https://github.com/apple/swift/pull/18081/files#diff-d7d8a9d1d74c516efb2594723443b62dR294)

jrose-apple commented:
Thanks, I'll review this soon. Answers to your questions:

1. PRs should always go to the master branch first. We're getting to the point where we need to not be taking arbitrary changes on 4.2, so this might not make it there.

2. Probably in this case you'd want to use [UnqualifiedLookup](https://github.com/apple/swift/blob/master/include/swift/AST/NameLookup.h#L105-L110). But that can totally be a follow-up patch.

3. I think it's fine not to worry about that for now and just fall back to the Swift name. That can be more follow-up work if you want to try later. (I think the general strategy there would be to do a lookup for the class first, followed by a lookup for the other method. But you'd probably only want to print that in a message, not the "renamed" attribute.)

varokas commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -771,18 +771,18 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
       os << " SWIFT_WARN_UNUSED_RESULT";
     }
-
+    
```
nit: preserve whitespaces?

varokas commented on test/PrintAsObjC/availability.swift:
```diff
@@ -82,7 +106,7 @@
 @objc class Availability {
     @objc func alwaysAvailable() {}
-
+    
```
nit: perverse whitespaces?

varokas commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -798,13 +798,20 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+              printEncodedString(renamedObjCRuntimeName, false);
+            } else {
+              printEncodedString(AvAttr->Rename, false);
+            }
```
nit: This is identical to line 923-931, and almost the same as 839-847. consider refactor this into a function?

jrose-apple replied:
I'm with varokas on this. I think you could push this all into `findRenamedDecl` and rename it to something like `printRenameForDecl`.

pitiphong-p commented:
@jrose-apple thank you. I’ll work on the number 2 on tomorrow and May try the number 3 too

pitiphong-p committed:
Reserve the white spaces

pitiphong-p committed:
Use the custom @objc name in the Availablity attributes on classes and protocols when generates the ObjC header

pitiphong-p committed:
Added: Test for the class and protocol cases for the PrintAsObjC availability

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     visitPart(RST->getReferentType(), optionalKind);
   }
   
+  Optional<ValueDecl *> findRenamedDecl(const swift::AvailableAttr *AvAttr, const swift::Decl *D) {
```
Nitpick: We have an 80-character column limit, but you can also drop the `swift::` namespace whenever you're in a Swift source file.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    Optional<ValueDecl *>renamedFuncDecl = None;
+
+    if (isa<ClassDecl>(D) || isa<ProtocolDecl>(D)) {
+      UnqualifiedLookup lookup(renamedDeclName.getBaseIdentifier(), declContext->getModuleScopeContext(), nullptr);
```
Nitpick: You can pass `UnqualifiedLookup::Flags::TypeLookup` to the "options" parameter here to make the lookup slightly more efficient.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(D->getASTContext());
+    
+    auto declContext = D->getDeclContext();
+    Optional<ValueDecl *>renamedFuncDecl = None;
```
Nitpick: This isn't always a function now. :-) You can also probably use `nullptr` as your "not found" value and drop the `Optional`. It's a little more subtle, but a pretty common idiom in our C++ code.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    if (!isa<ValueDecl>(D))
+      return None;
+    
+    SmallVector<ValueDecl *, 4> lookupResults;
```
Nitpick: Might as well move this into the qualified lookup branch now!

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        if (candidate->getKind() == DeclKind::Func && isa<FuncDecl>(candidate)
+            && cast<FuncDecl>(candidate)->getParameterLists().size() !=
+            cast<FuncDecl>(D)->getParameterLists().size())
+          continue;
```
I think you're trying to check that the number of parameters is the same here? That's not a bad idea, but it's not `getParameterLists()` that you want. Fortunately Slava P just added a useful API for what you do want: `getParameters()`.

Another good thing to check is whether they're both instance members.

pitiphong-p replied:
THAT IS GREAT!!! Thank you Slava P :D

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+        if (candidate->getKind() != D->getKind())
+          continue;
+        
+        if (candidate->getKind() == DeclKind::Func && isa<FuncDecl>(candidate)
```
It seems like you can use `isa` for `candidate` too, no?

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+                                   renamedDeclName, NL_QualifiedDefault, NULL,
+                                   lookupResults);
+      for (auto candidate : lookupResults) {
+        if (!shouldInclude(candidate))
```
Good idea here—easy way to filter for ObjC-compatible things only!

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -1884,6 +1906,46 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          continue;
+        
+        if (renamedFuncDecl)
+        /* TODO: Diagnose to compiler to tell the user that we found multiple candidates */;
```
I think it's okay to just fail in this case, at least for now. Incremental improvements are good!

pitiphong-p replied:
How should I make it fail here? Just fallback to the name in the rename attribute?

jrose-apple replied:
Yeah, that's what it's doing now. We could also consider not putting *anything* in the "rename" attribute when we can't get an ObjC name, and rely on the message. But for now, let's keep the existing behavior as our fallback.

jrose-apple reviewed (commented):
This is looking great! I have a lot of small comments, but the general approach is good.

I'd also like to see a lot more tests:
- Renames where there are multiple methods with the same full name but different parameter types.
- Renames where the renamed-to method comes from Objective-C.
- Renames where the renamed-to method is a class method instead of an instance method or vice versa.
- Renames of initializers.
- Renames of methods to properties or properties to methods.
- Renames to things that just don't exist.

…and that's all I can think of for now. But worth testing, yeah? :-)

Thanks again for picking this up!

pitiphong-p committed:
Pass the `UnqualifiedLookup::Flags::TypeLookup` option when doing the Type lookup

pitiphong-p committed:
Ditch the unneccessary Optional type on renamedFuncDecl variable

pitiphong-p committed:
Rename findRenamedDecl method to printRenameForDecl

pitiphong-p committed:
Added: Test case for the renamed to one of the overload methods case

pitiphong-p committed:
Added: Test case for the renamed to one of the initializer case

pitiphong-p committed:
Added: Test case for the case where the renamed method is class method not instance method

pitiphong-p committed:
Added: Test case for the case where the renamed method is class method not instance method and vice versa

pitiphong-p committed:
Fix the wrong checking condition on the case where both of them should be instance members or otherwise

pitiphong-p commented:
@jrose-apple I fixed and updated the code as you suggested. Also I added some test cases too

- [x] Renames where there are multiple methods with the same full name but different parameter types.
- [ ] Renames where the renamed-to method comes from Objective-C.
- [x] Renames where the renamed-to method is a class method instead of an instance method or vice versa.
- [x] Renames of initializers.
- [ ] Renames of methods to properties or properties to methods.
- [x] Renames to things that just don't exist.

However I do have a few questions.
1. `Renames where the renamed-to method comes from Objective-C.` So I need to have another .h file to test this, right?
2. `Renames of methods to properties or properties to methods` I'm not sure what's the expected behavior here. Should the finding fail and fall back to the old behavior?

Please have a look again :)

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -917,6 +915,60 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+  void printRenameForDecl(const AvailableAttr *AvAttr, const Decl *D,
+                          bool includeQuotes) {
+    if (AvAttr->Rename.empty() && isa<ValueDecl>(D))
+      return ;
```
We've already checked that this is a ValueDecl at all call sites, so you can probably just drop that. (Also, stray space before the semicolon.)

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -917,6 +915,60 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+          continue;
+        
+        if (renamedFuncDecl)
+        /* TODO: Diagnose to compiler to tell the user that we found multiple candidates */;
```
Still gotta do *something* here before we check this in. Just keeping track of a second boolean flag is probably fine.

pitiphong-p replied:
So we should just fail here and fallback to the `AvAttr->Rename` here?

jrose-apple replied:
Yeah, the same thing we'd do for a "we found nothing that matches" case.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -917,6 +915,60 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(D->getASTContext());
+    
+    auto declContext = D->getDeclContext();
+    ValueDecl *renamedFuncDecl = nullptr;
```
Still not necessarily a function anymore. :-) Also, may as well mark this `const`.

pitiphong-p replied:
`Still not necessarily a function anymore` I'm not sure what you mean here

pitiphong-p replied:
`may as well mark this const` but I need to change the renamedFuncDecl pointer after this. Or I misunderstood your point?

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -45,11 +45,29 @@
 // CHECK-DAG: SWIFT_AVAILABILITY(ios_app_extension,unavailable)
 // CHECK-DAG: SWIFT_AVAILABILITY(tvos_app_extension,unavailable)
 // CHECK-DAG: SWIFT_AVAILABILITY(watchos_app_extension,unavailable)
+// CHECK-DAG: ;
```
I think it's okay to leave this out, but if you want to have it you should use `CHECK-SAME` rather than `CHECK-DAG`. The "DAG" suffix means it's reorderable with other `CHECK-DAG` lines immediately before and after this one.

pitiphong-p replied:
Oh I just learn about `CHECK-SAME`, thanks

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -45,11 +45,29 @@
+// CHECK-NEXT: + (void)makeDeprecatedAvailabilityWithValue:(NSInteger)value SWIFT_DEPRECATED_MSG("use something else", "deprecatedAvailabilityWithValue:");
+// CHECK-NEXT: + (void)unavailableAvailabilityWithValue:(NSInteger)value;
+// CHECK-NEXT: + (void)makeUnavailableAvailabilityWithValue:(NSInteger)value
+// CHECK-DAG: SWIFT_UNAVAILABLE_MSG("'__makeUnavailableAvailability' has been renamed to 'unavailableAvailabilityWithValue:': use something else");
```
CHECK-SAME would be a better choice here too (since in this case there's only one thing to match).

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -148,71 +200,169 @@
+}
+
+@objc(SWTReplacementAvailableProtocol) protocol ReplacementAvailableProtocol {
+  @objc(methodReplacingInDeprecatedClassWithFirst:second:) func methodReplacingInDeprecatedClass(first: Int, second: Int) -> Void
```
I'll note that this is the same as the default name; you may want to pick something else.

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -148,71 +200,169 @@
+    @objc func deprecatedMethodRenamedToOverloadMethod1(first: Int, second: Int) {}
+  
+    @objc(overloadMethodWithFirst:second:) func overloadMethod1(first: Int, second: Int) {}
+    @objc func overloadMethod2(first: Double, second: Double) {}
```
Wait, these aren't overloads if you give them different base names.

pitiphong-p replied:
OH what a mistake!!!

jrose-apple commented:
Oh, *more* possible test cases:

- renaming to a method that does *not* have a custom Objective-C name (but is still exposed to Objective-C)
- renaming to a method that is not exposed to Objective-C (maybe with an explicit `@nonobjc`)

pitiphong-p committed:
Mark the `renamedFuncDecl` variable as `const`

pitiphong-p committed:
Change the type of D parameter to ValueDecl

pitiphong-p committed:
Use the `CHECK-SAME` where appropriate and update some ObjC names

pitiphong-p committed:
Added: renaming to a method that is not exposed to Objective-C

pitiphong-p committed:
Added: Renames of deprecated methods to properties or properties to methods and vice versa

pitiphong-p committed:
Fallback to the name in the Attribute when found duplicated candidates

pitiphong-p committed:
Added: Test case for the renamed to one of the overload methods which have a differnce ObjC names case

pitiphong-p committed:
Added: renaming to a method that does not have a custom Objective-C name test case

pitiphong-p commented:
- [x] Renames where there are multiple methods with the same full name but different parameter types.
- [x] Renames where the renamed-to method is a class method instead of an instance method or vice versa.
- [x] Renames of initializers.
- [x] Renames of methods to properties or properties to methods.
- [x] Renames to things that just don't exist.
- [x] Renaming to a method that is not exposed to Objective-C (maybe with an explicit @nonobjc)
- [x] Renaming to a method that does not have a custom Objective-C name (but is still exposed to Objective-C)
- [ ] Renames where the renamed-to method comes from Objective-C.

I added more test cases as you suggested. Could you review it again?

However there is an error while I was trying to add a test case for `Renames where the renamed-to method comes from Objective-C` case. Here's a standard error output

> Exit Code: 134
> 
> Command Output (stderr):
> --
> 0  swiftc                   0x000000010e28f968 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
> 1  swiftc                   0x000000010e28eba7 llvm::sys::RunSignalHandlers() + 39
> 2  swiftc                   0x000000010e28ffe2 SignalHandler(int) + 258
> 3  libsystem_platform.dylib 0x00007fff5c583f5a _sigtramp + 26
> 4  libsystem_platform.dylib 0x0000000120dae314 _sigtramp + 3296895956
> 5  libsystem_c.dylib        0x00007fff5c3211ae abort + 127
> 6  swiftc                   0x000000010b9d61df swift::ModuleFile::readConformance(llvm::BitstreamCursor&, swift::GenericEnvironment*) + 2207
> 7  swiftc                   0x000000010b9d5e6e swift::ModuleFile::readConformance(llvm::BitstreamCursor&, swift::GenericEnvironment*) + 1326
> 8  swiftc                   0x000000010b9ecc9d swift::ModuleFile::loadAllConformances(swift::Decl const*, unsigned long long, llvm::SmallVectorImpl<swift::ProtocolConformance*>&) + 237
> 9  swiftc                   0x000000010bc4be11 swift::ConformanceLookupTable::updateLookupTable(swift::NominalTypeDecl*, swift::ConformanceLookupTable::ConformanceStage, swift::LazyResolver*) + 193
> 10 swiftc                   0x000000010bc4c651 swift::ConformanceLookupTable::updateLookupTable(swift::NominalTypeDecl*, swift::ConformanceLookupTable::ConformanceStage, swift::LazyResolver*) + 2305
> 11 swiftc                   0x000000010bc4c1a5 swift::ConformanceLookupTable::updateLookupTable(swift::NominalTypeDecl*, swift::ConformanceLookupTable::ConformanceStage, swift::LazyResolver*) + 1109
> 12 swiftc                   0x000000010bc4fa37 swift::ConformanceLookupTable::lookupConformances(swift::NominalTypeDecl*, swift::DeclContext*, swift::LazyResolver*, swift::ConformanceLookupKind, llvm::SmallVectorImpl<swift::ProtocolDecl*>*, llvm::SmallVectorImpl<swift::ProtocolConformance*>*, llvm::SmallVectorImpl<swift::ConformanceDiagnostic>*) + 55
> 13 swiftc                   0x000000010bd0f323 swift::DeclContext::getLocalProtocols(swift::ConformanceLookupKind, llvm::SmallVectorImpl<swift::ConformanceDiagnostic>*, bool) const + 131
> 14 swiftc                   0x000000010b5d21dd swift::printAsObjC(llvm::raw_ostream&, swift::ModuleDecl*, llvm::StringRef, swift::AccessLevel) + 3965
> 15 swiftc                   0x000000010aa64fe2 printAsObjCIfNeeded(llvm::StringRef, swift::ModuleDecl*, llvm::StringRef, bool) + 450
> 16 swiftc                   0x000000010aa60d0a performCompile(swift::CompilerInstance&, swift::CompilerInvocation&, llvm::ArrayRef<char const*>, int&, swift::FrontendObserver*, swift::UnifiedStatsReporter*) + 7146
> 17 swiftc                   0x000000010aa5e0af swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2943
> 18 swiftc                   0x000000010aa18048 main + 1128
> 19 libdyld.dylib            0x00007fff5c275015 start + 1
> 20 libdyld.dylib            0x0000000000000017 start + 2748887043
> Stack dump:
> 0.	Program arguments: /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swiftc -frontend -target x86_64-apple-macosx10.9 -module-cache-path /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -swift-version 4 -enable-source-import -sdk /Swift-Repo/swift/test/Inputs/clang-importer-sdk -I /Swift-Repo/swift/test/Inputs/clang-importer-sdk/swift-modules -parse-as-library /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-macosx-x86_64/PrintAsObjC/Output/availability.swift.tmp/availability.swiftmodule -typecheck -emit-objc-header-path /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-macosx-x86_64/PrintAsObjC/Output/availability.swift.tmp/availability.h -import-objc-header /Swift-Repo/swift/test/PrintAsObjC/../Inputs/empty.h -disable-objc-attr-requires-foundation-module 
> 1.	While generating Objective-C header
> 2.	While loading conformances for 'SubClassOfBaseClass' in module 'availability'
> 3.	While reading inherited conformance for type 'SubClassOfBaseClass'
> 4.	While cross-referencing conformance for 'NSObject' in module 'ObjectiveC'
> 5.	While ... to 'Equatable' in module 'Swift'
> 6.	If you're seeing a crash here, check that your SDK and dependencies are at least as new as the versions used to build 'availability'
> /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-macosx-x86_64/PrintAsObjC/Output/availability.swift.script: line 5: 51065 Abort trap: 6           /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swiftc -frontend -target x86_64-apple-macosx10.9 -module-cache-path '/Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/swift-test-results/x86_64-apple-macosx10.9/clang-module-cache' -sdk '/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk' -swift-version 4 -enable-source-import -sdk '/Swift-Repo/swift/test/Inputs/clang-importer-sdk' -I '/Swift-Repo/swift/test/Inputs/clang-importer-sdk/swift-modules' -parse-as-library /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-macosx-x86_64/PrintAsObjC/Output/availability.swift.tmp/availability.swiftmodule -typecheck -emit-objc-header-path /Swift-Repo/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/test-macosx-x86_64/PrintAsObjC/Output/availability.swift.tmp/availability.h -import-objc-header /Swift-Repo/swift/test/PrintAsObjC/../Inputs/empty.h -disable-objc-attr-requires-foundation-module
>

What I did was adding `import Foundation` and a new subclass of `Coat` class in the `availability.swift`

How could I fix this?

pitiphong-p commented:
@jrose-apple I also implement the type following feature (e.g. `renamed: "AnotherClass.thisIsAMethod(_:)"`) feature. You can have a look here https://github.com/pitiphong-p/swift/compare/objc-rename-attribute...pitiphong-p:objc-rename-attribute-type-following

Should I include this change with this PR?

jrose-apple commented:
> What I did was adding `import Foundation` and a new subclass of `Coat` class in the availability.swift
> 
> How could I fix this?

Most of the PrintAsObjC tests, including availability.swift, use the "mock SDK" in test/Inputs/clang-importer-sdk/ in order to build faster, as well as decreasing the likelihood that a test will need to be updated for a new Xcode release. The catch, though, is that when you use the mock SDK, you have to set up appropriate equivalents for the overlays as well. You can see this in test/PrintAsObjC/classes.swift as "-enable-source-import hackaround".

There are a few options here: you can copy that workaround over to availability.swift, or you could make a new test file that uses the real SDK instead of the mock SDK (which would mostly just mean removing the "(mock-sdk: %clang-importer-sdk)" and "-disable-objc-attr-requires-foundation-module" from the `RUN` lines in availability.swift). Either one's fine with me.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -917,6 +915,62 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(D->getASTContext());
+    
+    auto declContext = D->getDeclContext();
+    const ValueDecl *renamedFuncDecl = nullptr;
```
This still shouldn't be called "renamedFuncDecl" when it's very often not a function. :-)

pitiphong-p replied:
Oh right, sorry that I missed that

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -917,6 +915,62 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+  void printRenameForDecl(const AvailableAttr *AvAttr, const ValueDecl *D,
+                          bool includeQuotes) {
+    if (AvAttr->Rename.empty())
+      return;
```
I think you can assert this rather than early-exiting. Any caller would already have to deal with the no-rename case differently.

(Sorry I didn't make this comment last time. I guess I was distracted by the formatting.)

pitiphong-p replied:
OK, I'll change this to assert instead

jrose-apple reviewed (commented):
The code logic looks good, just a few tiny more changes I suggest. I'll look over all the tests I asked you to add tomorrow (Pacific time).

jrose-apple commented:
> @jrose-apple I also implement the type following feature (e.g. `renamed: "AnotherClass.thisIsAMethod(_:)"`) feature. You can have a look here pitiphong-p/swift@objc-rename-attribute...pitiphong-p:objc-rename-attribute-type-following
> 
> Should I include this change with this PR?

At this point I feel bad for dragging *this* one out, so let's not include that. That way we can merge this PR sooner.

pitiphong-p committed:
Rename renamedFuncDecl to renamedDecl

pitiphong-p committed:
Change from early exit to assert instead

pitiphong-p commented:
>At this point I feel bad for dragging this one out, so let's not include that. That way we can merge this PR sooner.

OK I'll open it as a new PR since I also have some ideas on improving this further :)

pitiphong-p committed:
Added: New test file for the renaming to the Objective-C functions

pitiphong-p committed:
Add the deprecation messages in the renamed-to Objective-C name test cases

pitiphong-p commented:
@jrose-apple I fixed the code and added test cases you asked for yesterday already. Please have a look again.
However I only added the deprecation test cases but will added the unavailable test cases after you finish checking the deprecated test cases.

P.S. I also added a new `availability-real-sdk.swift` for the renamed-to ObjC name test cases. I chose the `removing the "(mock-sdk: %clang-importer-sdk)" and "-disable-objc-attr-requires-foundation-module" ` approach since another approach caused a build failure while running the test

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -45,13 +45,40 @@
+// CHECK-NEXT: + (void)makeDeprecatedAvailabilityWithValue:(NSInteger)value SWIFT_DEPRECATED_MSG("use something else", "deprecatedAvailabilityWithValue:");
+// CHECK-NEXT: + (void)unavailableAvailabilityWithValue:(NSInteger)value;
+// CHECK-NEXT: + (void)makeUnavailableAvailabilityWithValue:(NSInteger)value
+// CHECK-SAME: SWIFT_UNAVAILABLE_MSG("'__makeUnavailableAvailability' has been renamed to 'unavailableAvailabilityWithValue:': use something else");
```
Hm, yeah, we should indeed fix the unavailable macro.

pitiphong-p replied:
Fix the unavailable macro? How?

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -64,6 +91,40 @@
+// CHECK-NEXT: - (nonnull instancetype)initWithNewZ:(NSInteger)z OBJC_DESIGNATED_INITIALIZER;
+// CHECK-NEXT: @end
+
+// CHECK-LABEL: SWIFT_AVAILABILITY(macos,deprecated=0.0.1,message="'DeprecatedAvailability' has been renamed to 'SWTReplacementAvailable'")
```
Another case where we ought to support `renamed=""` but don't. (Not a problem for this PR.)

jrose-apple reviewed (commented):
Tests all look good! Thanks for ticking all the boxes I asked for.

jrose-apple commented:
@swift-ci Please test

pitiphong-p commented:
@jrose-apple Thank you. I'll add the test cases for the unavailable in the afternoon Thailand time and will ask Swift-CI to run the test. Hope it'll be ready to merge within this week :D

pitiphong-p committed:
Added: Test cases for the platform deprecation, unavailable and platform unavailable PrintAsObjC

pitiphong-p commented:
@swift-ci please test

pitiphong-p commented:
@swift-ci Please test

pitiphong-p commented:
@jrose-apple I added the remaining platform deprecation, unavailable and platform unavailable test cases. Could you have a look again and ask the @swift-ci to run a test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on test/PrintAsObjC/availability.swift:
```diff
@@ -242,29 +354,61 @@
     @objc public func methodWithoutCustomObjCName(value: Int) -> Int { return -1 }
 
+    @available(*, deprecated, renamed: "methodWithoutCustomObjCName(value:)")
+    @objc public func deprecatedMethodRenamedToMethodWithouCustomObjCName(value: Int) -> Int { return -1 }
```
Not that it's super important, but for this family you wrote "Withou" instead of "Without".

pitiphong-p replied:
@jrose-apple Should I fix it here or can leave it for the next PR?

jrose-apple replied:
I think it's okay to leave it. Let's get this merged!

jrose-apple commented:
Thanks once again for working on this! Looking forward to more patches from you in the future.

pitiphong-p commented:
@jrose-apple OMG OMG OMG Thank you for helping me from the start. I'm glad to have a chance to work on this and on Swift open source project. I'm looking forward to contribute to Swift in the future too :D
